markdown_it/rules_core/linkify.py

```diff
@@ -0,0 +1,130 @@
+import re
+
+from ..common.utils import arrayReplaceAt
+from ..common.normalize_url import normalizeLinkText, normalizeLink, validateLink
+from .state_core import StateCore
+from ..token import Token
+
+
+LINK_OPEN_RE = re.compile(r"^<a[>\s]", flags=re.IGNORECASE)
+LINK_CLOSE_RE = re.compile(r"^</a\s*>", flags=re.IGNORECASE)
+
+HTTP_RE = re.compile(r"^http://")
+MAILTO_RE = re.compile(r"^mailto:")
+TEST_MAILTO_RE = re.compile(r"^mailto:", flags=re.IGNORECASE)
+
+
+def isLinkOpen(string: str):
+    return LINK_OPEN_RE.search(string)
+
+
+def isLinkClose(string: str):
+    return LINK_CLOSE_RE.search(string)
+
+
+def linkify(state: StateCore):
+    blockTokens = state.tokens
+
+    if not state.md.options.linkify:
+        return
+
+    if not state.md.linkify:
+        raise Exception("Linkify enabed but not installed.")
+
+    for j in range(len(blockTokens)):
+        if blockTokens[j].type != "inline" or not state.md.linkify.pretest(
+            blockTokens[j].content
+        ):
+            continue
+
+        tokens = blockTokens[j].children
+
+        htmlLinkLevel = 0
+
+        # We scan from the end, to keep position when new tags added.
+        # Use reversed logic in links start/end match
+        assert tokens is not None
+        for i in range(len(tokens))[::-1]:
```

Comment:
This is just a preference thing but I personally always found `reversed(iterable)` more readable than `iterable[::-1]`

Reply:
`for` statement did not calculate the index `i` correctly.

So I changed to a `while` statement.